### PR TITLE
Add `with` to signal_stage

### LINKED ISSUE #116: `signal_superseded()` not using with=
`signal_superseded()` drops `with=` and `when=` when using `signal_stage()` 

From https://github.com/tidyverse/dplyr/pull/6037#pullrequestreview-770564955

### PR DESCRIPTION
**Pull request by hadley**
Fixes #116

### CODE REVIEW

File: tests/testthat/_snaps/signal.md [hunk 1]
```diff
@@ -9,3 +9,16 @@
     Output
       <lifecycle_stage: foo(arg) is superseded>
 
+# signal_stage supports with
+
+    Code
+      (expect_condition(signal_stage("superseded", "foo()", "bar()")))
+    Output
+      <lifecycle_stage: foo() is superseded
```
Comment (lionel-) [inline tests/testthat/_snaps/signal.md:17]: Should we add a `testthat_print()` method for bare conditions to make the display less awkward? hmm I guess that's not really worth the effort.

File: tests/testthat/test-signal.R [hunk 2]
```diff
@@ -6,7 +6,7 @@ test_that("signal stage captures desired data", {
   cnd <- expect_condition(f(), class = "lifecycle_stage")
   expect_equal(cnd$stage, "experimental")
   expect_equal(cnd$package, "pkg")
-  expect_equal(cnd$"function", "foo")
+  expect_equal(cnd$function_nm, "foo")
```
Comment (lionel-) [inline tests/testthat/test-signal.R:9]: oops did we partial match here?

**Review by lionel- — APPROVED** [review]

File: tests/testthat/test-signal.R (hunk 2 above)
Comment (hadley) [reply tests/testthat/test-signal.R:9]: Yup 😑

File: tests/testthat/_snaps/signal.md (hunk 1 above)
Comment (hadley) [reply tests/testthat/_snaps/signal.md:17]: Yeah, and it can also be resolved by #91, which would give other folks an easy way to see these annotations.

**Pull request merged**